File: citrus_press_factory.py
def citrus_press_factory(press_type):
    """Creates a citrus press."""
    press_types = {"lime", "lemon", "orange", "grapefruit"}

    if not press_type in press_types:
        raise ValueError(f"press_type must be one of {press_type}")

    compatible_fruit_types_lookup = {
        "lime": {"lime", "lemon"},
        "lemon": {"lime", "lemon"},
        "orange": {"lemon", "orange"},
        "grapefruit": {"grapefruit", "pomelo"},
    }

    def citrus_press(fruit):
        """Makes fresh juice from a fruit."""
        compatible_fruit_types = compatible_fruit_types_lookup[press_type]

        print(compatible_fruit_types)

        if not fruit in compatible_fruit_types:
            raise ValueError(f"fruit must be one of {compatible_fruit_types}")

        return f"Some fresh {fruit} juice."

    return citrus_press


if __name__ == "__main__":

    lemon_press = citrus_press_factory("lemon")
    print(lemon_press("lime"))
    print(lemon_press("lemon"))
    print(lemon_press("orange"))  # should throw an error

    pomelo_press = citrus_press_factory("pomelo")  # should throw an error

File: citrus_press_factory.js
/**
 * Creates a citrus press.
 */
const citrusPressFactory = (pressType) => {
  const pressTypes = new Set(["lime", "lemon", "orange", "grapefruit"]);

  if (!pressTypes.has(pressType)) {
    throw Error(
      `pressType must be one of ${Array.from(pressTypes).join(", ")}).`
    );
  }

  compatibleFruitTypesLookup = {
    lime: new Set(["lime", "lemon"]),
    lemon: new Set(["lime", "lemon"]),
    orange: new Set(["lemon", "orange"]),
    grapefruit: new Set(["grapefruit", "pomelo"]),
  };

  /**
   * Makes fresh juice from a fruit.
   */
  const citrusPress = (fruit) => {
    const compatibleFruitTypes = compatibleFruitTypesLookup[pressType];

    if (!compatibleFruitTypes.has(fruit)) {
      throw Error(
        `fruit must be one of ${Array.from(compatibleFruitTypes).join(", ")}.`
      );
    }

    return `Some fresh ${fruit} juice`;
  };

  return citrusPress;
};

const lemonPress = citrusPressFactory("lemon");
console.log(lemonPress("lime"));
console.log(lemonPress("lemon"));
console.log(lemonPress("orange")); // should throw an error

const pomeloPress = citrusPressFactory("pomelo"); // should throw an error
